File: packages/core/src/theme.ts
/**
 * IntentText Theme System — v2.10
 *
 * Themes are JSON-based design value sets applied by the renderer.
 * The .it format never changes. Themes are purely a renderer concern.
 * Built-in themes ship as JSON files in src/themes/.
 */

import corporateJson from "./themes/corporate.json";
import minimalJson from "./themes/minimal.json";
import warmJson from "./themes/warm.json";
import technicalJson from "./themes/technical.json";
import printJson from "./themes/print.json";
import legalJson from "./themes/legal.json";
import editorialJson from "./themes/editorial.json";
import darkJson from "./themes/dark.json";

export interface ThemeFonts {
  body: string;
  heading: string;
  mono: string;
  size: string;
  leading: string;
}

export interface ThemeColors {
  text: string;
  heading: string;
  muted: string;
  accent: string;
  border: string;
  background: string;
  "code-bg": string;
  "trust-approved"?: string;
  "trust-signed"?: string;
  "trust-frozen"?: string;
  "trust-warning"?: string;
  watermark?: string;
  [key: string]: string | undefined;
}

export interface ThemeSpacing {
  "page-margin": string;
  "section-gap": string;
  "block-gap": string;
  indent: string;
}

export interface ThemePrint {
  "header-font-size"?: string;
  "footer-font-size"?: string;
  "header-color"?: string;
  "footer-color"?: string;
}

export interface IntentTheme {
  name: string;
  version: string;
  description?: string;
  author?: string;
  fonts: ThemeFonts;
  colors: ThemeColors;
  spacing: ThemeSpacing;
  blocks?: Record<string, Record<string, string | boolean>>;
  print?: ThemePrint;
}

/** Map of built-in theme names to their JSON data. */
const BUILTIN_THEMES: Record<string, IntentTheme> = {};

/** Register a built-in theme (called at module load). */
export function registerBuiltinTheme(theme: IntentTheme): void {
  BUILTIN_THEMES[theme.name] = theme;
}

/** Get a built-in theme by name. Returns undefined if not found. */
export function getBuiltinTheme(name: string): IntentTheme | undefined {
  return BUILTIN_THEMES[name];
}

/** List all built-in theme names. */
export function listBuiltinThemes(): string[] {
  return Object.keys(BUILTIN_THEMES);
}

/** Resolve a color value — if it matches a theme color key, return that color. */
function resolveColor(value: string, colors: ThemeColors): string {
  if (value in colors) return colors[value] ?? value;
  return value;
}

/** Font stack helper: append generic fallback families. */
function fontStack(font: string, type: "body" | "heading" | "mono"): string {
  if (type === "mono") return `'${font}', monospace`;
  return `${font}, system-ui, sans-serif`;
}

/**
 * Generate CSS custom properties and block styles from a theme.
 * Mode controls which additional styles to include.
 */
export function generateThemeCSS(
  theme: IntentTheme,
  mode: "web" | "print" = "web",
): string {
  const c = theme.colors;
  const f = theme.fonts;
  const s = theme.spacing;
  const p = theme.print;

  // CSS custom properties
  let css = `:root{`;
  css += `--it-font-body:${fontStack(f.body, "body")};`;
  css += `--it-font-heading:${fontStack(f.heading, "heading")};`;
  css += `--it-font-mono:${fontStack(f.mono, "mono")};`;
  css += `--it-font-size:${f.size};`;
  css += `--it-leading:${f.leading};`;
  for (const [key, val] of Object.entries(c)) {
    if (val !== undefined) css += `--it-color-${key}:${val};`;
  }
  css += `--it-spacing-page-margin:${s["page-margin"]};`;
  css += `--it-spacing-section-gap:${s["section-gap"]};`;
  css += `--it-spacing-block-gap:${s["block-gap"]};`;
  css += `--it-spacing-indent:${s.indent};`;
  css += `}\n`;

  // Apply custom properties to the document
  css += `.intent-document{font-family:var(--it-font-body);font-size:var(--it-font-size);line-height:var(--it-leading);color:var(--it-color-text);background:var(--it-color-background);}\n`;
  css += `.intent-title,.intent-section,.intent-sub{font-family:var(--it-font-heading);color:var(--it-color-heading);}\n`;
  css += `.intent-section{margin-top:var(--it-spacing-section-gap);}\n`;
  css += `.intent-note,.intent-prose{margin-bottom:var(--it-spacing-block-gap);}\n`;
  css += `code,.intent-code{font-family:var(--it-font-mono);background:var(--it-color-code-bg);}\n`;
  css += `.intent-summary{border-left-color:var(--it-color-muted);color:var(--it-color-muted);}\n`;
  css += `.intent-divider-line{border-color:var(--it-color-border);}\n`;

  // Block-level styles from theme
  if (theme.blocks) {
    for (const [blockType, styles] of Object.entries(theme.blocks)) {
      const selector = `.intent-${blockType}`;
      const declarations: string[] = [];
      for (const [prop, val] of Object.entries(styles)) {
        if (typeof val === "boolean") {
          if (val && prop === "border-bottom")
            declarations.push(`border-bottom:1px solid var(--it-color-border)`);
        } else {
          // Resolve color references
          const resolved = resolveColor(val, c);
          declarations.push(`${prop}:${resolved}`);
        }
      }
      if (declarations.length > 0) {
        css += `${selector}{${declarations.join(";")}}\n`;
      }
    }
  }

  // Print-specific styles
  if (mode === "print" && p) {
    const hColor = p["header-color"] ? resolveColor(p["header-color"], c) : "";
    const fColor = p["footer-color"] ? resolveColor(p["footer-color"], c) : "";
    if (p["header-font-size"] || hColor) {
      let decl = "";
      if (p["header-font-size"]) decl += `font-size:${p["header-font-size"]};`;
      if (hColor) decl += `color:${hColor};`;
      css += `@page{@top-left{${decl}}@top-center{${decl}}@top-right{${decl}}}\n`;
    }
    if (p["footer-font-size"] || fColor) {
      let decl = "";
      if (p["footer-font-size"]) decl += `font-size:${p["footer-font-size"]};`;
      if (fColor) decl += `color:${fColor};`;
      css += `@page{@bottom-left{${decl}}@bottom-center{${decl}}@bottom-right{${decl}}}\n`;
    }
  }

  return css;
}

// ── Register built-in themes from JSON files ────────────

registerBuiltinTheme(corporateJson as IntentTheme);
registerBuiltinTheme(minimalJson as IntentTheme);
registerBuiltinTheme(warmJson as IntentTheme);
registerBuiltinTheme(technicalJson as IntentTheme);
registerBuiltinTheme(printJson as IntentTheme);
registerBuiltinTheme(legalJson as IntentTheme);
registerBuiltinTheme(editorialJson as IntentTheme);
registerBuiltinTheme(darkJson as IntentTheme);

File: packages/rust/src/theme.rs
//! Theme system parity surface.

use std::collections::HashMap;
use std::sync::RwLock;

use once_cell::sync::Lazy;

#[derive(Debug, Clone, Default)]
pub struct ThemeFonts {
    pub body: String,
    pub heading: String,
    pub mono: String,
    pub size: String,
    pub leading: String,
}

#[derive(Debug, Clone, Default)]
pub struct ThemeSpacing {
    pub page_margin: String,
    pub section_gap: String,
    pub block_gap: String,
    pub indent: String,
}

#[derive(Debug, Clone, Default)]
pub struct IntentTheme {
    pub name: String,
    pub version: String,
    pub description: Option<String>,
    pub author: Option<String>,
    pub fonts: ThemeFonts,
    pub colors: HashMap<String, String>,
    pub spacing: ThemeSpacing,
    pub blocks: HashMap<String, HashMap<String, String>>,
}

static BUILTIN_THEMES: Lazy<RwLock<HashMap<String, IntentTheme>>> = Lazy::new(|| {
    let mut themes = HashMap::new();
    themes.insert("minimal".to_string(), minimal_theme());
    themes.insert("corporate".to_string(), corporate_theme());
    themes.insert("dark".to_string(), dark_theme());
    RwLock::new(themes)
});

pub fn register_builtin_theme(theme: IntentTheme) {
    if let Ok(mut map) = BUILTIN_THEMES.write() {
        map.insert(theme.name.clone(), theme);
    }
}

pub fn get_builtin_theme(name: &str) -> Option<IntentTheme> {
    BUILTIN_THEMES.read().ok()?.get(name).cloned()
}

pub fn list_builtin_themes() -> Vec<String> {
    let mut names: Vec<String> = BUILTIN_THEMES
        .read()
        .map(|m| m.keys().cloned().collect())
        .unwrap_or_default();
    names.sort();
    names
}

pub fn generate_theme_css(theme: &IntentTheme, mode: Option<&str>) -> String {
    let mode = mode.unwrap_or("web");
    let mut css = String::from(":root{");
    css.push_str(&format!("--it-font-body:{};", theme.fonts.body));
    css.push_str(&format!("--it-font-heading:{};", theme.fonts.heading));
    css.push_str(&format!("--it-font-mono:{};", theme.fonts.mono));
    css.push_str(&format!("--it-font-size:{};", theme.fonts.size));
    css.push_str(&format!("--it-leading:{};", theme.fonts.leading));
    for (k, v) in &theme.colors {
        css.push_str(&format!("--it-color-{k}:{v};"));
    }
    css.push_str("}\n");

    css.push_str(".intent-document{font-family:var(--it-font-body);font-size:var(--it-font-size);line-height:var(--it-leading);color:var(--it-color-text);background:var(--it-color-background);}\n");
    css.push_str(".intent-title,.intent-section,.intent-sub{font-family:var(--it-font-heading);color:var(--it-color-heading);}\n");
    css.push_str("code,.intent-code{font-family:var(--it-font-mono);background:var(--it-color-code-bg);}\n");

    for (block, styles) in &theme.blocks {
        css.push_str(&format!(".intent-{block}{{"));
        for (prop, val) in styles {
            css.push_str(&format!("{prop}:{val};"));
        }
        css.push_str("}\n");
    }

    if mode == "print" {
        css.push_str("@media print{.intent-document{margin:0;}}\n");
    }
    css
}

fn base_theme(name: &str, body: &str, heading: &str, bg: &str, text: &str) -> IntentTheme {
    let mut colors = HashMap::new();
    colors.insert("text".to_string(), text.to_string());
    colors.insert("heading".to_string(), heading.to_string());
    colors.insert("muted".to_string(), "#6b7280".to_string());
    colors.insert("accent".to_string(), "#2563eb".to_string());
    colors.insert("border".to_string(), "#d1d5db".to_string());
    colors.insert("background".to_string(), bg.to_string());
    colors.insert("code-bg".to_string(), "#f3f4f6".to_string());

    IntentTheme {
        name: name.to_string(),
        version: "1.0".to_string(),
        description: None,
        author: Some("intenttext-rust".to_string()),
        fonts: ThemeFonts {
            body: body.to_string(),
            heading: heading.to_string(),
            mono: "ui-monospace, SFMono-Regular, Menlo, monospace".to_string(),
            size: "16px".to_string(),
            leading: "1.6".to_string(),
        },
        colors,
        spacing: ThemeSpacing {
            page_margin: "24px".to_string(),
            section_gap: "28px".to_string(),
            block_gap: "14px".to_string(),
            indent: "20px".to_string(),
        },
        blocks: HashMap::new(),
    }
}

fn minimal_theme() -> IntentTheme {
    base_theme("minimal", "system-ui, sans-serif", "system-ui, sans-serif", "#ffffff", "#111827")
}

fn corporate_theme() -> IntentTheme {
    base_theme("corporate", "Inter, system-ui, sans-serif", "Inter, system-ui, sans-serif", "#ffffff", "#1f2937")
}

fn dark_theme() -> IntentTheme {
    let mut t = base_theme("dark", "system-ui, sans-serif", "system-ui, sans-serif", "#0f172a", "#e5e7eb");
    t.colors.insert("code-bg".to_string(), "#111827".to_string());
    t
}
